lib/rspec/core/configuration.rb

```diff
@@ -202,10 +202,31 @@ def only_failures_but_not_configured?
         only_failures? && !example_status_persistence_file_path
       end
 
-      # @macro add_setting
+      # @macro define_reader
       # If specified, indicates the number of failures required before cleaning
-      # up and exit (default: `nil`).
-      add_setting :fail_fast
+      # up and exit (default: `nil`). Can also be `true` to fail on exit on first
+      # failure
+      define_reader :fail_fast
+
+      # @see fail_fast
+      def fail_fast=(value)
+        case value
+        when true, 'true'
+          @fail_fast = true
+        when false, 'false'
+          @fail_fast = false
+        when nil
+          @fail_fast = nil
+        else
+          if value.to_i == 0
+            raise ArgumentError, "Cannot set `RSpec.configuration.fail_fast`" \
```

Comment:
I decided to throw here so that the users can know when they set this value incorrectly, and not have to wonder why the configuration is not working.
Would this then require a major version bump?

Do you prefer not to have this config throw?

EDIT: We could probably just issue a warning?

Reply:
I'm on the fence, if the previous behaviour is considered truthy then a warning should be issued and it considered truthy, with a note to deprecate and raise in RSpec 4, but otherwise I consider this a bug fix and it can be raised.

Reply:
🤔

Thinking more about this, I'd lean towards warning + raise in rspec 4, so that users don't get failing rspec runner, for a non-major version update.

I'll go ahead and change to warning. I think it will create less frustration than raising an error.

Reply:
I found three different ways the warnings are being set in this file:

```ruby
warn "RSpec's reporter has already been initialized with " \
  "#{deprecation_stream.inspect} as the deprecation stream, so your change to "\
  "`deprecation_stream` will be ignored. You should configure it earlier for " \
  "it to take effect, or use the `--deprecation-out` CLI option. " \
  "(Called from #{CallerFilter.first_non_rspec_line})"

RSpec.warning "Configuring `#{name}` to #{value} has no effect since " \
              "RSpec has already loaded the spec files."

RSpec.warn_with "WARNING: `:suite` hooks do not support metadata since " \
                "they apply to the suite as a whole rather than " \
                "any individual example or example group that has metadata. " \
                "The metadata you have provided (#{meta.inspect}) will be ignored."
```

I went for using `warn`.

Reply:
Can you use one of our helpers please, not Rubys bare warn. It provides a configurable suppression point.

Reply:
Which one would you recommend? I'll switch to `RSpec.warning`

Reply:
Either is fine, warning adds "WARNING" to the front if you don't want to type it yourself :joy:

Reply:
Just to note, the original behavior here is an error raised in rspec code due to this line:
https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/reporter.rb#L227

```
Failure/Error: fail_fast <= @failed_examples.size

ArgumentError:
  comparison of String with 1 failed
```

This makes the test fail, but continues executing the following tests. The following tests all fail, no matter whether they would pass in isolation (because `config.fail_fast` gets checked in reporter on every test I guess).